### PR TITLE
feat(zk): add create proof method to prover

### PR DESCRIPTION
**Pull request by dongchangYoo**
In this PR,`CreateProof()` method was added to `Prover` at `tachyon/zk/plonk/halo2/prover.h`. And It includes several bug fixes and a bit of refactoring for the existing logics. See [CreateProof()](https://github.com/kroma-network/halo2/blob/7d0a36990452c8e7ebd600de258420781a9b7917/halo2_proofs/src/plonk/prover.rs#L41-L682).

From now on, `Tachyon` can generate a halo2-proofs.

### CODE REVIEW

File: tachyon/zk/plonk/halo2/argument.h [hunk 1]
```diff
@@ -129,23 +129,40 @@ class Argument {
...
-                             } else {
-                               for (size_t i = 0; i < prover->pcs().N(); ++i) {
-                                 CHECK(prover->GetWriter()->WriteToTranscript(
-                                     *instance_column[i]));
-                               }
-                             }
-                             return prover->domain()->IFFT(instance_column);
-                           });
-                     });
+      const std::vector<std::vector<std::vector<F>>>& instance_3d_vec) {
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/argument.h:300]: Is there a reason to name `instance_3d_vec`? If not, please name it `instance_columns_vec` for consistency with other places.

File: tachyon/zk/plonk/halo2/argument.h [hunk 2]
```diff
@@ -129,23 +129,40 @@ class Argument {
...
+    std::vector<std::vector<Poly>> instance_polys_vec;
+    instance_polys_vec.reserve(num_circuit);
+    for (size_t i = 0; i < num_circuit; ++i) {
+      const std::vector<std::vector<F>>& instance_columns = instance_3d_vec[i];
+      std::vector<Poly> instance_polys;
+      instance_polys.reserve(num_instance_columns);
+      for (size_t j = 0; j < num_instance_columns; ++j) {
+        const std::vector<F>& instances = instance_columns[j];
+        if constexpr (PCS::kQueryInstance) {
+          Evals instance_column(instances);
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/argument.h:318]: It seems that this line should be executed for both cases.

**Comment by chokobole** [comment]
Please change the type of https://github.com/kroma-network/tachyon/pull/238/commits/0d7faea006009772904a63665128ac59b75df1d8 to `refac`

File: tachyon/zk/plonk/halo2/synthesizer.h [hunk 3]
```diff
@@ -89,14 +89,30 @@ class Synthesizer {
     }
   }
 
-  // Return |challenge_| as a vector.
+  // Return |challenge_| as a vector instead of |btree_map|.
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/synthesizer.h:94]: This doesn't seem to be related to the commit title "feat(zk): add takers of Synthesizer"

File: tachyon/zk/plonk/halo2/synthesizer.h [hunk 4]
```diff
@@ -89,14 +89,30 @@ class Synthesizer {
...
   std::vector<F> ExportChallenges() {
     return base::Map(
         std::make_move_iterator(challenges_.begin()),
         std::make_move_iterator(challenges_.end()),
         [](std::pair<size_t, F>&& item) { return std::move(item.second); });
   }
 
+  // Return |challenge_| as a vector instead of |btree_map| and then clear it.
+  std::vector<F> TakeChallenges() {
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/synthesizer.h:103]: For me, `TakeXXX()` always has to be with `&&`.

File: tachyon/zk/plonk/halo2/point_set.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+template <typename Point>
+class PointSet {
+ public:
+  PointSet() = default;
+  explicit PointSet(size_t expected_size) { points_.reserve(expected_size); }
+
+  const std::vector<std::unique_ptr<Point>>& points() const { return points_; }
+
+  // Insert
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/point_set.h:26]: Please remove a trivial comment

File: tachyon/zk/plonk/circuit/examples/simple_v1_circuit_unittest.cc [hunk 6]
```diff
@@ -630,6 +630,35 @@ TEST_F(SimpleV1CircuitTest, LoadProvingKey) {
...
+  ProvingKey<PCS> pkey;
+  ASSERT_TRUE(pkey.Load(prover_.get(), circuit));
+  ASSERT_TRUE(
+      prover_->CreateProof(pkey, std::move(instance_columns_vec), circuits));
+
+  std::vector<uint8_t> proof = prover_->GetWriter()->buffer().owned_buffer();
+  std::vector<uint8_t> expected_proof(std::begin(kExpectedProof),
+                                      std::end(kExpectedProof));
+  ASSERT_EQ(proof.size(), 1024);
+  ASSERT_EQ(proof, expected_proof);
```
Comment (chokobole) [inline tachyon/zk/plonk/circuit/examples/simple_v1_circuit_unittest.cc:659]: Please use https://google.github.io/googletest/reference/matchers.html#container-matchers

File: tachyon/zk/plonk/halo2/prover.h [hunk 7]
```diff
@@ -62,6 +66,73 @@ class Prover : public ProverBase<PCS> {
     return ret;
   }
 
+  template <typename Circuit>
+  [[nodiscard]] bool CreateProof(
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/prover.h:70]: Is there a case that it returns false?

File: tachyon/zk/plonk/halo2/argument.h [hunk 8]
```diff
@@ -129,23 +129,40 @@ class Argument {
...
-                                 CHECK(prover->GetWriter()->WriteToTranscript(
-                                     *instance_column[i]));
-                               }
-                             }
-                             return prover->domain()->IFFT(instance_column);
-                           });
-                     });
+      const std::vector<std::vector<std::vector<F>>>& instance_3d_vec) {
+    CHECK_GT(instance_3d_vec.size(), size_t{0});
+    size_t num_circuit = instance_3d_vec.size();
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/halo2/argument.h:302]:
```suggestion
    size_t num_circuit = instance_3d_vec.size();
    CHECK_GT(num_circuit, size_t{0});
```

File: tachyon/zk/plonk/halo2/argument.h [hunk 9]
```diff
@@ -129,23 +129,40 @@ class Argument {
...
+
+    std::vector<std::vector<Poly>> instance_polys_vec;
+    instance_polys_vec.reserve(num_circuit);
+    for (size_t i = 0; i < num_circuit; ++i) {
+      const std::vector<std::vector<F>>& instance_columns = instance_3d_vec[i];
+      std::vector<Poly> instance_polys;
+      instance_polys.reserve(num_instance_columns);
+      for (size_t j = 0; j < num_instance_columns; ++j) {
+        const std::vector<F>& instances = instance_columns[j];
+        if constexpr (PCS::kQueryInstance) {
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/halo2/argument.h:317]:
```suggestion
        if constexpr (PCS::kSupportsBatchMode && PCS::kQueryInstance) {
```

File: tachyon/zk/plonk/halo2/synthesizer.h [hunk 10]
```diff
@@ -89,14 +89,30 @@ class Synthesizer {
...
+    std::vector<F> challenges = base::Map(
+        std::make_move_iterator(challenges_.begin()),
+        std::make_move_iterator(challenges_.end()),
+        [](std::pair<size_t, F>&& item) { return std::move(item.second); });
+    challenges_.clear();
+    return challenges;
+  }
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/halo2/synthesizer.h:110]:
```suggestion
  // Move out |challenges_| as a vector instead of |btree_map|.
  std::vector<F>&& TakeChallenges() && {
      return base::Map(
        std::make_move_iterator(challenges_.begin()),
        std::make_move_iterator(challenges_.end()),
        [](std::pair<size_t, F>&& item) { return std::move(item.second); });
  }
```
I'm not sure this code will compile but I thought we don't need to clear `challenges_` since we're moving it out.

**Comment by TomTaehoonKim** [comment]
@dongchangYoo Could you fix [1eec8e3](https://github.com/kroma-network/tachyon/pull/238/commits/1eec8e3cee857294c96528dc745c7c09d94eaf82) as a test commit?

File: tachyon/zk/plonk/halo2/point_set.h [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        base::FindIndexIf(points_.begin(), points_.end(),
+                          [&item](const std::unique_ptr<Point>& point) {
+                            return *point == item;
+                          });
+    Point* point_ptr = nullptr;
+    if (existing_idx.has_value()) {
+      point_ptr = points_[existing_idx.value()].get();
+    } else {
+      points_.push_back(std::make_unique<Point>(item));
+      point_ptr = points_[points_.size() - 1].get();
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/halo2/point_set.h:38]:
```suggestion
      point_ptr = points_.back().get();
```

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

File: tachyon/zk/plonk/halo2/point_set.h (hunk 11 above)
Comment (chokobole) [reply tachyon/zk/plonk/halo2/point_set.h:38]: I was considering this, too! I didn't comment about this because I thought it's just a personal preference difference.

File: tachyon/zk/plonk/halo2/BUILD.bazel [hunk 12]
```diff
@@ -126,12 +137,12 @@ tachyon_cc_library(
     testonly = True,
     hdrs = ["prover_test.h"],
     deps = [
+        ":blake2b_transcript",
         ":prover",
         "//tachyon/math/elliptic_curves/bn/bn254",
         "//tachyon/math/polynomials/univariate:univariate_evaluation_domain_factory",
         "//tachyon/zk/base:blinder",
         "//tachyon/zk/base/commitments:shplonk_extension",
-        "//tachyon/zk/plonk/halo2:blake2b_transcript",
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/BUILD.bazel:134]: Could you separate the change from this commit even if it's a small change?

File: tachyon/zk/plonk/halo2/BUILD.bazel [hunk 13]
```diff
@@ -175,21 +197,41 @@ tachyon_cc_library(
...
     ],
     deps = [
+        ":argument",
         ":blake2b_transcript",
         ":pinned_verifying_key",
         ":poseidon_transcript",
         ":prover_test",
         ":sha256_transcript",
+        ":synthesizer",
+        "//tachyon/base/containers:container_util",
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/BUILD.bazel:268]: Please remove this

File: tachyon/zk/plonk/halo2/argument.h (hunk 9 above)
Comment (chokobole) [reply tachyon/zk/plonk/halo2/argument.h:317]: I think this is more correct.

```suggestion
        if constexpr (PCS::kQueryInstance) {
          if constexpr (PCS::kSupportsBatchMode) {
            // BatchCommit
          } else {
            // Commit
          }
        }         
```

File: tachyon/zk/plonk/halo2/argument.h [hunk 14]
```diff
@@ -129,23 +129,40 @@ class Argument {
...
+          Evals instance_column(instances);
+          prover->BatchCommitAt(instance_column, i * num_instance_columns + j);
+          instance_polys.push_back(prover->domain()->IFFT(instance_column));
+        } else {
+          for (size_t i = 0; i < instances.size(); ++i) {
+            CHECK(prover->GetWriter()->WriteToTranscript(instances[i]));
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/argument.h:323]: You use `i`, `j` and `i`. So I want you to use range-based for loop or use `k`

File: tachyon/zk/plonk/halo2/argument.h (hunk 2 above)
Comment (chokobole) [reply tachyon/zk/plonk/halo2/argument.h:318]: Plus, leave a comment something like "FIXME(dongchangYoo): This causes copy internally and I want to avoid this"

**Comment by chokobole** [comment]
Could you make changes into each commit in  https://github.com/kroma-network/tachyon/pull/238/commits/a42815498a3b4523b9ad22d984d99f94183444ba ?

File: tachyon/zk/plonk/halo2/argument.h [hunk 15]
```diff
@@ -171,19 +163,18 @@ class Argument {
   // not owned
   const std::vector<Poly>* fixed_polys_ = nullptr;
 
-  std::vector<std::vector<Evals>> advice_columns_vec_;
-  std::vector<std::vector<F>> advice_blinds_vec_;
   // Note(dongchangYoo): to optimize memory usage, release every advice
   // evaluations after generating an advice polynomial. That is, when
   // |advice_transformed_| is set to true, |advice_values_by_circuits| is
   // released, and only |advice_polys_by_circuits| becomes available for use.
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/argument.h]: Do you need to update this comment? I can't find `advice_values_by_circuits`

**Comment by chokobole** [comment]
Please remove the tests at the beginning not in the middle. https://github.com/kroma-network/tachyon/pull/238/commits/1eec8e3cee857294c96528dc745c7c09d94eaf82

File: tachyon/zk/plonk/halo2/point_set.h
Comment (chokobole) [inline tachyon/zk/plonk/halo2/point_set.h:1]: I believe this file should be placed in `zk/base` or `crypto`.

File: tachyon/zk/plonk/halo2/point_set.h [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+    }
+    return base::DeepRef<const Point>(point_ptr);
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/point_set.h:40]:
```suggestion
    if (!existing_idx.has_value()) {
      existing_idx = points_.size();
      points_.push_back(std::make_unique<Point>(item));
    }
    return base::DeepRef<const Point>(points_[existing_idx.value()].get());
```

File: tachyon/zk/lookup/lookup_argument_runner.h [hunk 17]
```diff
@@ -11,13 +11,15 @@
 
 #include "gtest/gtest_prod.h"
 
+#include "tachyon/crypto/commitments/polynomial_openings.h"
 #include "tachyon/zk/base/entities/prover_base.h"
 #include "tachyon/zk/base/prover_query.h"
```
Comment (chokobole) [inline tachyon/zk/lookup/lookup_argument_runner.h:16]: Please remove `ProverQuery` file, too then

File: tachyon/zk/plonk/permutation/permutation_argument_runner_impl.h [hunk 18]
```diff
@@ -111,57 +113,49 @@ PermutationArgumentRunner<Poly, Evals>::EvaluateCommitted(
...
+    ret.emplace_back(base::DeepRef<const Poly>(&poly), x_next_ref,
+                     poly.Evaluate(x_next));
   }
 
-  int32_t blinding_factors =
-      static_cast<int32_t>(prover->blinder().blinding_factors());
-  F x_last = Rotation(-(blinding_factors + 1)).RotateOmega(prover->domain(), x);
+  base::CheckedNumeric<int32_t> blinding_factors =
+      prover->blinder().blinding_factors();
+  Rotation last_rotation = Rotation((-blinding_factors - 1).ValueOrDie());
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/permutation_argument_runner_impl.h:143]: This change should be separated and also shouldn't be changed in this way. There's other cases that use this pattern as well.

**Comment by chokobole** [comment]
Could you separate the changes https://github.com/kroma-network/tachyon/pull/238/commits/2297be7e2dbde4148f41ce6babc3d5a453c5c88c. At least, there's a change which is not related to bugfix.

File: tachyon/zk/plonk/halo2/synthesizer.h [hunk 19]
```diff
@@ -58,16 +63,12 @@ class Synthesizer {
...
-        }
-        for (size_t j = 0; j < phases.size(); ++j) {
-          if (current_phase != phases[j]) continue;
+        std::vector<Phase> advice_phases =
+            constraint_system_->advice_column_phases();
+        for (size_t j = 0; j < rational_advice_columns.size(); ++j) {
+          if (current_phase != advice_phases[j]) continue;
           const RationalEvals& column = rational_advice_columns[j];
-          std::vector<F> evaluated;
+          std::vector<F> evaluated(column.NumElements());
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/synthesizer.h:71]: This change should be separated

File: tachyon/zk/plonk/halo2/BUILD.bazel [hunk 20]
```diff
@@ -240,5 +241,6 @@ tachyon_cc_unittest(
         "//tachyon/base/containers:container_util",
         "//tachyon/zk/plonk/circuit/examples:simple_circuit",
         "//tachyon/zk/plonk/circuit/floor_planner:simple_floor_planner",
+        "//tachyon/zk/plonk/keys:proving_key",
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/BUILD.bazel:271]: This change should be separated.

File: tachyon/zk/plonk/permutation/permutation_argument_runner_impl.h [hunk 21]
```diff
@@ -38,6 +38,9 @@ PermutationArgumentRunner<Poly, Evals>::CommitArgument(
   // will never underflow because of the requirement of at least a degree
   // 3 circuit for the permutation argument.
   CHECK_GE(constraint_system_degree, argument.RequiredDegree());
+  if (argument.columns().empty()) {
+    return {};
+  }
```
Comment (Insun35) [inline tachyon/zk/plonk/permutation/permutation_argument_runner_impl.h:43]: Please remove the curly braces for consistency with the code below.
```suggestion
  if (argument.columns().empty()) return {};
```

File: tachyon/zk/plonk/halo2/point_set.h (hunk 11 above)
Comment (TomTaehoonKim) [reply tachyon/zk/plonk/halo2/point_set.h]: @dongchangYoo You may resolve this comment if you think change is not needed!

File: tachyon/zk/plonk/halo2/synthesizer.h [hunk 22]
```diff
@@ -97,6 +97,20 @@ class Synthesizer {
         [](std::pair<size_t, F>&& item) { return std::move(item.second); });
   }
 
+  // Return |challenge_| as a vector instead of |btree_map| and then clear it.
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/halo2/synthesizer.h:102]:
```suggestion
  // Move out |challenge_| as a vector.
```

File: tachyon/zk/plonk/halo2/argument.h (hunk 9 above)
Comment (TomTaehoonKim) [reply tachyon/zk/plonk/halo2/argument.h:318]: @chokobole I've checked it. Please resolve this comment after you check as well!

**Comment by TomTaehoonKim** [comment]
@dongchangYoo How about fixing [09d21d5](https://github.com/kroma-network/tachyon/pull/238/commits/09d21d5eac40bc1c98e3f1978eb9a04f07a79e7d) as a test commit?

File: tachyon/zk/plonk/halo2/argument_util.h [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+      });
+}
+
+template <typename PCS, typename Poly, typename Evals, typename F>
+std::vector<PermutationCommitted<Poly>> BatchCommitPermutations(
+    ProverBase<PCS>* prover, const PermutationArgument& permutation_argument,
+    const PermutationProvingKey<Poly, Evals>& permutation_proving_key,
+    const std::vector<RefTable<Evals>>& tables, size_t cs_degree, const F& beta,
+    const F& gamma) {
+  size_t num_circuits = tables.size();
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/halo2/argument_util.h:91]: Could you add "stddef.h" for `size_t`?

File: tachyon/zk/plonk/halo2/argument.h [hunk 24]
```diff
@@ -202,7 +352,7 @@ class Argument {
   // not owned
   const std::vector<Poly>* fixed_polys_ = nullptr;
 
-  // Note(dongchangYoo): to optimize memory usage, release every advice
+  // NOTE(dongchangYoo): to optimize memory usage, release every advice
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/halo2/argument.h:355]: Could you make this change as a separate chore commit please?

File: tachyon/zk/plonk/halo2/argument.h [hunk 25]
```diff
@@ -125,6 +126,155 @@ class Argument {
...
+
+    return {std::move(evaluated_permutations), std::move(evaluated_lookups_vec),
+            std::move(evaluated_vanishing)};
+  }
+
+  template <typename P, typename L, typename V>
+  std::vector<crypto::PolynomialOpening<Poly>> ConstructOpenings(
+      ProverBase<PCS>* prover, const ProvingKey<PCS>& proving_key,
+      const StepReturns<P, L, V>& evaluated, const F& x,
+      PointSet<F>& opening_points_set_) {
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/halo2/argument.h:213]:
```suggestion
      PointSet<F>& opening_points_set) {
```

File: tachyon/zk/plonk/halo2/argument.h [hunk 26]
```diff
@@ -125,6 +126,155 @@ class Argument {
...
+    std::vector<RefTable<Poly>> tables = ExportPolyTables();
+    const ConstraintSystem<F>& cs =
+        proving_key.verifying_key().constraint_system();
+    opening_points_set_.Insert(x);
+
+    std::vector<crypto::PolynomialOpening<Poly>> ret;
+    ret.reserve(GetNumOpenings(proving_key, evaluated, num_circuits_));
+    for (size_t i = 0; i < num_circuits_; ++i) {
+      // Generate openings for instance of the specific circuit.
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/halo2/argument.h:222]:
```suggestion
      // Generate openings for instances of the specific circuit.
```

File: tachyon/zk/plonk/halo2/argument.h [hunk 27]
```diff
@@ -125,6 +126,155 @@ class Argument {
...
+      if constexpr (PCS::kQueryInstance) {
+        std::vector<crypto::PolynomialOpening<Poly>> openings =
+            GenerateColumnOpenings(prover, tables[i].instance_columns(),
+                                   cs.instance_queries(), x,
+                                   opening_points_set_);
+        ret.insert(ret.end(), std::make_move_iterator(openings.begin()),
+                   std::make_move_iterator(openings.end()));
+      }
+
+      // Generate openings for advice of the specific circuit.
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/halo2/argument.h:232]:
```suggestion
      // Generate openings for advices of the specific circuit.
```

File: tachyon/zk/plonk/halo2/argument.h [hunk 28]
```diff
@@ -125,6 +126,155 @@ class Argument {
...
+      std::vector<crypto::PolynomialOpening<Poly>> openings =
+          GenerateColumnOpenings(prover, tables[i].advice_columns(),
+                                 cs.advice_queries(), x, opening_points_set_);
+      ret.insert(ret.end(), std::make_move_iterator(openings.begin()),
+                 std::make_move_iterator(openings.end()));
+
+      // Generate openings for permutation of the specific circuit.
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/halo2/argument.h:239]:
```suggestion
      // Generate openings for permutations of the specific circuit.
```

File: tachyon/zk/plonk/halo2/argument.h [hunk 29]
```diff
@@ -125,6 +126,155 @@ class Argument {
...
+          [prover, &x, &opening_points_set_](
+              const LookupEvaluated<Poly>& evaluated_lookup) {
+            return LookupArgumentRunner<Poly, Evals>::OpenEvaluated(
+                prover, evaluated_lookup, x, opening_points_set_);
+          });
+      ret.insert(ret.end(), std::make_move_iterator(openings.begin()),
+                 std::make_move_iterator(openings.end()));
+    }
+
+    // Generate openings for fixed column.
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/halo2/argument.h:257]:
```suggestion
    // Generate openings for fixeds
```
Or for comments for advices and lookups, please add columns for them as well for consistency.

File: tachyon/zk/plonk/halo2/prover.h [hunk 30]
```diff
@@ -62,6 +66,73 @@ class Prover : public ProverBase<PCS> {
...
+            proving_key.permutation_proving_key(),
+            std::move(permuted_lookups_vec), beta, gamma);
+
+    F y = writer->SqueezeChallenge();
+    argument.TransformAdvice(this->domain());
+    ExtendedEvals circuit_evals = argument.GenerateCircuitPolynomial(
+        this, proving_key, committed_result, beta, gamma, theta, y);
+    VanishingConstructed<PCS> constructed_vanishing;
+    CHECK(CommitFinalHPoly(this, std::move(committed_result).TakeVanishing(),
+                           proving_key.verifying_key(), circuit_evals,
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/halo2/prover.h:118]: Sorry but could you rename parameter name of `circuit_column` in `CommitFinalHPoly` to circuit_evals?

File: tachyon/zk/plonk/circuit/examples/simple_circuit_unittest.cc [hunk 31]
```diff
@@ -714,7 +744,8 @@ TEST_F(SimpleCircuitTest, Verify) {
   {
     expected_vanishing_random_poly_commitment = CreateCommitment(
         {"0x0000000000000000000000000000000000000000000000000000000000000001",
-         "0x0000000000000000000000000000000000000000000000000000000000000002"});
+         "0x000000000000000000000000000000000000000000000000000000000000000"
+         "2"});
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/examples/simple_circuit_unittest.cc:748]:
```suggestion
         "0x0000000000000000000000000000000000000000000000000000000000000002"});
```
In my PC, this works! I don't think this change is necessary!

File: tachyon/zk/plonk/circuit/examples/simple_circuit_unittest.cc [hunk 32]
```diff
@@ -727,9 +758,11 @@ TEST_F(SimpleCircuitTest, Verify) {
   {
     std::vector<Point> points = {
         {"0x2157c5e9888d3770a99534af98e5256e87a8156013477212abaf2dcb61267cc1",
-         "0x2b04d3cec50250e35b5919da1cfe8142db0a636c0f29c6427fe713f52c1fff60"},
+         "0x2b04d3cec50250e35b5919da1cfe8142db0a636c0f29c6427fe713f52c1fff6"
+         "0"},
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/examples/simple_circuit_unittest.cc:762]:
```suggestion
         "0x2b04d3cec50250e35b5919da1cfe8142db0a636c0f29c6427fe713f52c1fff60"},
```

File: tachyon/zk/plonk/circuit/examples/simple_circuit_unittest.cc [hunk 33]
```diff
@@ -727,9 +758,11 @@ TEST_F(SimpleCircuitTest, Verify) {
...
         {"0x1f74e7de13dea2916e273fdd6dc8fb6b439880af6f5a51f43c1519fff52df80c",
-         "0x1fd85823b8d30be9fdde823da1f8c4c876bb874a236050c3d04032999ac56eb7"},
+         "0x1fd85823b8d30be9fdde823da1f8c4c876bb874a236050c3d04032999ac56eb"
+         "7"},
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/examples/simple_circuit_unittest.cc:765]:
```suggestion
         "0x1fd85823b8d30be9fdde823da1f8c4c876bb874a236050c3d04032999ac56eb7"},
```

File: tachyon/zk/plonk/circuit/examples/simple_lookup_circuit.h [hunk 34]
```diff
@@ -25,6 +25,10 @@ class SimpleLookupConfig {
                      const AdviceColumnKey& advice)
       : selector_(selector), table_(table), advice_(advice) {}
 
+  SimpleLookupConfig Clone() const {
+    return SimpleLookupConfig(selector_, table_, advice_);
+  }
+
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/examples/simple_lookup_circuit.h:31]: Please remove this.

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

File: tachyon/zk/plonk/circuit/examples/simple_lookup_circuit.h (hunk 34 above)
Comment (dongchangYoo) [reply tachyon/zk/plonk/circuit/examples/simple_lookup_circuit.h:31]: A circuit config class should have `Clone()`, since `FloorPlanner` consume a config of the circuit.
And this change will be made in a separate, exclusive commit!

**Comment by TomTaehoonKim** [comment]
@dongchangYoo Can you fix the commit message of 90c482b to "perf(zk): avoid"? I think current commit message is inappropriate.

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by Insun35 — APPROVED** [review]
LGTM

**Comment by chokobole** [comment]
I believe the type of https://github.com/kroma-network/tachyon/pull/238/commits/90c482b84f2f563f605e2bf05f4aff46a6e52e05 should be `perf`.

**Comment by chokobole** [comment]
Is [this](https://github.com/kroma-network/tachyon/pull/238/commits/1074506001943c5eb9e4b83004862a2f1474a918) a fix commit then ?

**Comment by dongchangYoo** [comment]
> I believe the type of [90c482b](https://github.com/kroma-network/tachyon/commit/90c482b84f2f563f605e2bf05f4aff46a6e52e05) should be `perf`.

This method acts like `InPlace`. That is, if `&` does not exist, this method is like it does nothing.
So, i think this commit's type is "fix".

File: tachyon/zk/plonk/halo2/argument_util.h [hunk 35]
```diff
@@ -0,0 +1,186 @@
...
+      num_openings += evaluated.permutations()[i].product_polys().size();
+    }
+    if (!evaluated.lookups_vec().empty()) {
+      num_openings += evaluated.lookups_vec()[i].size();
+    }
+  }
+  num_openings += proving_key.permutation_proving_key().permutations().size();
+
+  // Ensure that there is exactly one opening for vanishing.
+  return num_openings + 1;
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/argument_util.h:167]: Could you also apply `base::CheckedNumeric<>` here too?

File: tachyon/zk/base/blinder.h [hunk 36]
```diff
@@ -28,12 +28,17 @@ class Blinder {
...
 
-  // Blinds |evals| at behind by |blinding_factors_|.
-  // Returns false if |evals.NumElements()| is less than |blinding_factors_|.
-  bool Blind(Evals& evals) {
+  // The number of |blinding_rows| is determined to be either
+  // |blinding_factors_| or |blinding_factors_| + 1, depending on the
+  // |include_boundary_row| option.
+  // Blinds |evals| at behind by |blinding_rows|.
+  // Returns false if |evals.NumElements()| is less than |blinding_rows|.
+  bool Blind(Evals& evals, bool include_boundary_row = false) {
```
Comment (chokobole) [inline tachyon/zk/base/blinder.h:36]: Please rename `boundary` to `last` as per https://zcash.github.io/halo2/design/proving-system/permutation.html#zero-knowledge-adjustment

File: tachyon/zk/lookup/permute_expression_pair.h [hunk 37]
```diff
@@ -134,8 +134,8 @@ template <typename PCS, typename Evals, typename F = typename Evals::Field>
   Evals input(std::move(permuted_input_expressions));
   Evals table(std::move(permuted_table_expressions));
 
-  prover->blinder().Blind(input);
-  prover->blinder().Blind(table);
+  prover->blinder().Blind(input, /*include_boundary_row*/ true);
+  prover->blinder().Blind(table, /*include_boundary_row*/ true);
```
Comment (chokobole) [inline tachyon/zk/lookup/permute_expression_pair.h:138]:
```suggestion
  prover->blinder().Blind(input, /*include_boundary_row=*/ true);
  prover->blinder().Blind(table, /*include_boundary_row=*/ true);
```

**Comment by chokobole** [comment]
https://github.com/kroma-network/tachyon/pull/238/commits/a7ecd38272399448c860a0098c4ef2dabba7a0c3 has a typo `Fllor`

File: tachyon/zk/plonk/circuit/examples/simple_circuit_unittest.cc [hunk 38]
```diff
@@ -631,6 +632,35 @@ TEST_F(SimpleCircuitTest, LoadProvingKey) {
...
+      std::move(instance_columns)};
+
+  ProvingKey<PCS> pkey;
+  ASSERT_TRUE(pkey.Load(prover_.get(), circuit));
+  prover_->CreateProof(pkey, std::move(instance_columns_vec), circuits);
+
+  std::vector<uint8_t> proof = prover_->GetWriter()->buffer().owned_buffer();
+  std::vector<uint8_t> expected_proof(std::begin(kExpectedProof),
+                                      std::end(kExpectedProof));
+  ASSERT_EQ(proof.size(), 1024);
```
Comment (chokobole) [inline tachyon/zk/plonk/circuit/examples/simple_circuit_unittest.cc:660]: Please remove this since below line checks size as well.

File: tachyon/zk/plonk/circuit/examples/simple_lookup_circuit_unittest.cc [hunk 39]
```diff
@@ -521,6 +522,30 @@ TEST_F(SimpleLookupCircuitTest, LoadProvingKey) {
...
+      std::move(instance_columns)};
+
+  ProvingKey<PCS> pkey;
+  ASSERT_TRUE(pkey.Load(prover_.get(), circuit));
+  prover_->CreateProof(pkey, std::move(instance_columns_vec), circuits);
+
+  std::vector<uint8_t> proof = prover_->GetWriter()->buffer().owned_buffer();
+  std::vector<uint8_t> expected_proof(std::begin(kExpectedProof),
+                                      std::end(kExpectedProof));
+  ASSERT_EQ(proof.size(), 640);
```
Comment (chokobole) [inline tachyon/zk/plonk/circuit/examples/simple_lookup_circuit_unittest.cc:545]: Please remove this since below line checks size as well.

File: tachyon/zk/plonk/circuit/examples/simple_lookup_v1_circuit_unittest.cc [hunk 40]
```diff
@@ -520,6 +521,30 @@ TEST_F(SimpleLookupV1CircuitTest, LoadProvingKey) {
...
+      std::move(instance_columns)};
+
+  ProvingKey<PCS> pkey;
+  ASSERT_TRUE(pkey.Load(prover_.get(), circuit));
+  prover_->CreateProof(pkey, std::move(instance_columns_vec), circuits);
+
+  std::vector<uint8_t> proof = prover_->GetWriter()->buffer().owned_buffer();
+  std::vector<uint8_t> expected_proof(std::begin(kExpectedProof),
+                                      std::end(kExpectedProof));
+  ASSERT_EQ(proof.size(), 640);
```
Comment (chokobole) [inline tachyon/zk/plonk/circuit/examples/simple_lookup_v1_circuit_unittest.cc:544]: Please remove this since below line checks size as well.

File: tachyon/zk/plonk/circuit/examples/simple_v1_circuit_unittest.cc [hunk 41]
```diff
@@ -630,6 +631,34 @@ TEST_F(SimpleV1CircuitTest, LoadProvingKey) {
...
+      std::move(instance_columns)};
+
+  ProvingKey<PCS> pkey;
+  ASSERT_TRUE(pkey.Load(prover_.get(), circuit));
+  prover_->CreateProof(pkey, std::move(instance_columns_vec), circuits);
+
+  std::vector<uint8_t> proof = prover_->GetWriter()->buffer().owned_buffer();
+  std::vector<uint8_t> expected_proof(std::begin(kExpectedProof),
+                                      std::end(kExpectedProof));
+  ASSERT_EQ(proof.size(), 1024);
```
Comment (chokobole) [inline tachyon/zk/plonk/circuit/examples/simple_v1_circuit_unittest.cc:658]: Please remove this since below line checks size as well.

**Review by chokobole — APPROVED** [review]
LGTM
